### ISSUE TITLE
Missing Unitest coverage statistics

### ISSUE THREAD

**Issue by aviadavi** [rokucommunity/rooibos#44]
Hi,

It would be great to know the actual coverage of the unit-tests.
We want to increase our confidence in the application and alert when unit-tests below a specific percentage.

Thanks.

BR,
Aviad A

**Comment by georgejecook** [comment]
I have a pattern for doing this - I also have a preprocessor that I can take code from to simplify it.

Basically, I need to add a function to every scope, that allows us to report the code coverage stats, say it's called `RBS_CODE_COV(location`, and it returns _true_

This line of code needs to get injected into every single file, with special care for if statements, array declarations and aa declarations.

i.e.

```'@Test valid data
function PMT_parsePlayLIstSupplement_valid()
  RBS_CODECOV("#RBS_POS#"): json = TU_loadStubFile("playListSupplement", true)
  RBS_CODECOV("#RBS_POS#"): data = parsePlaylistSupplement(json)
  RBS_CODECOV("#RBS_POS#"): m.assertSubType(data, "TabContent")
  RBS_CODECOV("#RBS_POS#"): m.assertNodeCount(data, 16)
  RBS_CODECOV("#RBS_POS#"): item = data.getChild(0).getChild(0)
  RBS_CODECOV("#RBS_POS#"): m.assertSubType(item, "CustomContentNode")
  RBS_CODECOV("#RBS_POS#"): m.assertEqual(item.id, "id")
  RBS_CODECOV("#RBS_POS#"): m.assertEqual(item.title, "Happy Families")
  RBS_CODECOV("#RBS_POS#"): m.assertEqual(item.hdposterurl, "url.jpg")
  ' RBS_CODECOV("#RBS_POS#"): m.assertEqual(item.json.runtime, "43m")
  ' if RBS_CODECOV("#RBS_POS#") and m.foo = "bar"
  if RBS_CODECOV("#RBS_POS#") and m.foo = "bar"
  RBS_CODECOV("#RBS_POS#"):  ? "if test"
  end if
  RBS_CODECOV("#RBS_POS#"): for i = 0 to 10
  RBS_CODECOV("#RBS_POS#"): ? "for test"
  RBS_CODECOV("#RBS_POS#"): end for
  RBS_CODECOV("#RBS_POS#"): i = 0
  RBS_CODECOV("#RBS_POS#"): while i < 10
  RBS_CODECOV("#RBS_POS#"): i++
  RBS_CODECOV("#RBS_POS#"): ? "while test"
  RBS_CODECOV("#RBS_POS#"): end while
  RBS_CODECOV("#RBS_POS#"): m.obj = {
    a:1
    b:2
  }
  RBS_CODECOV("#RBS_POS#"): m.arr = [
    1,
    2,
    3,
  ]
end function```

If you want to start by contributing a new class called `CodeCoverageProcessor`, with a public function, which converts a string to inject these lines of code, as laid out above,  in the rooibosPreprocessor, then that will be a great help. I'll bring across my code from my other preprocessor that copies the entire source path, and has methods for getting source file text etc.

If not, I can make a basic version of this in a few weeks, which uses the brs language AST (which I leverage in my other project).

**Comment by georgejecook** [comment]
I'm just looking into whether I can get burp to do this - will be way easier. - in the meantime @aviadavi could you please tell me how you expect this feature to work? what metrics will it have?

**Comment by aviadavi** [comment]
HI,

Something like :
Total code lines: 10000
Covered lines: 7500
Coverage percentage: 75%

Same statistics per brs file would be great.

Function level statistics would be the awesome. 

Maybe an output file with all the covered lines marked?? This would be best.

**Comment by ronenbit1** [comment]
This would be great to have the coverage statistics.

**Comment by ronenlh** [comment]
Branch coverage might be better than line coverage and also easier to implement. Taken from [here](http://www.semdesigns.com/Company/Publications/TestCoverage.pdf).

To implement branch coverage, a processor would need to add code acting as a self-identifying probe in every scope according to brs syntax. This would look like an associative array with the location of the probe as key (maybe using the filename + the line number), and a boolean flag as its value

An initializing loop would reset all flags every time tests run, and a collector mechanism writes the associative array into an accumulator, which divides visited lines with all probes. This can also provide information on which part of the system has not been executed by tests.

@georgejecook is your preprocessor able to insert code in every scope?

**Comment by georgejecook** [comment]
Yes. I've already written what you've described.

I'll hopefully be finished this week depending on workload.

>

**Comment by georgejecook** [comment]
thanks for both your inputs, as per the rooibos 3.0.1 docs, I'm about to push up, here's a description of the coverage - I'd like to close this issue once you've verified the behaviour - if we want to track branching @ronenlh, can you please open a separate issue? thanks :)


### How coverage works

rooibos-preprocessor, via `rooibosC` command line tool, or the npm package will run against a root project folder, and using the `sourceFilePattern` globs your provide, will ascertain which files require coverage.

It then buils an Abstract Syntax Tree, using Sean Barag's wonderful [brs interpreter](https://github.com/sjbarag/brs), to ascertain which lines of the identified files can have their coverage tracked.

The preprocessor then injects function calls for all trackable lines, which update tracking variables at runtime.

Rooibos framework will then collate the stats and display them at the end of the run.

#### Statement support

The following statements types are supported:

  - variable assignments
  - method calls
  - nested function definitions (i.e. functions inside of arrays, variable assignments, method call args, or associative arrays)
  - if statement conditions
  - blocks of code in if, else if, and else statements
  - for and while loops
  - print statements

The following statements are _not_ supported

  - goto
  - named lines (e.g. `myGotoLine:`)
  - else if conditions (these are coming real soon though)

### Coverage report

When your coverage test run finishes rooibos will print out:

```
+++++++++++++++++++++++++++++++++++++++++++
Code Coverage Report
+++++++++++++++++++++++++++++++++++++++++++

Total Coverage: 5.066445% (610/12040)
Files:  17/ 100

HIT FILES
---------
pkg:/components/ContentRetriever/ContentRetriever.brs: 2.461539% (8/325)
pkg:/components/Analytics/AnalyticsManager.brs: 3.125% (6/192)
pkg:/components/Service/AuthenticationService.brs: 3.532609% (13/368)
pkg:/components/Container/UserContainer/UserContainer.brs: 3.703704% (1/27)
pkg:/components/MaintenanceCheckTask/AppConfigurationTask.brs: 7.407407% (2/27)
pkg:/components/Storage/PersistentStorage.brs: 15.27778% (11/72)
pkg:/source/Analytics/AnalyticsVideoMixin.brs: 16.92308% (22/130)
pkg:/components/DeveloperService/DeveloperService.brs: 33.33334% (2/6)
pkg:/components/Log/Log.brs: 45% (9/20)
pkg:/source/Modules/BaseModule.brs: 55.55556% (5/9)
pkg:/source/Log.brs: 65% (13/20)
pkg:/source/Modules/DeepLinkingModule.brs: 76.92308% (50/65)
pkg:/components/GlobalInitializer/GlobalInitializer.brs: 79.64601% (90/113)
pkg:/source/Modules/RecommendationModule.brs: 84.9315% (186/219)
pkg:/source/Mixins/ParsingMixin.brs: 86.47343% (179/207)
pkg:/source/Mixins/SMCErrors.brs: 92.30769% (12/13)
pkg:/source/Analytics/AnalyticsConstants.brs: 100% (1/1)

MISSED FILES
------------
pkg:/components/CustomRowListItem/CustomItemGenres/CustomItemGenres.brs: MISS!
pkg:/components/StringUtils.brs: MISS!
pkg:/components/Core/Components/TabComponent/TabComponent.brs: MISS!
pkg:/components/Core/Components/NoKeyPressRowList.brs: MISS!
pkg:/components/Model/TabComponentContent.brs: MISS!
```

  - Total coverage - % (num of hit lines/ num of trackable lines)
  - Files: num of hit files / total num of trackable files

Following is a list of all the hit files, and their coverage % and (hit lines/total lines)

Lastly the files that were not hit at all, during test execution.

The current implementation is capable of tracking lcov style statistics, and this will be implemented in the future.

**Comment by ronenlh** [comment]
As per my own tests the line coverage works great. I believe the issue can be closed. I'll open another one for branch coverage testing. Awesome work!

**Comment by georgejecook** [comment]
Closing this, as it's resolved in 3.0.0, which is in alpha right now; but should be on  mainline by next week.
